runtime/src/mock_channel.rs

```diff
@@ -0,0 +1,181 @@
+//! A mock implementation of a channel that implements the Sink and Stream traits.
+
+use crate::{Error, Sink, Stream};
+use bytes::Bytes;
+use futures::channel::oneshot;
+use std::{
+    collections::VecDeque,
+    sync::{Arc, Mutex},
+};
+
+// A mock channel struct that is used internally by the ByteSink and ByteStream.
+struct ByteChannel {
+    buffer: VecDeque<u8>,
+    waiter: Option<(usize, oneshot::Sender<Bytes>)>,
+}
+
+// Returns an async-safe sink/stream pair that share an underlying ByteChannel.
+// The buffer is wrapped in a mutex and an Arc to allow for shared ownership across async tasks.
+pub fn new() -> (ByteSink, ByteStream) {
+    let channel = Arc::new(Mutex::new(ByteChannel {
+        buffer: VecDeque::new(),
+        waiter: None,
+    }));
+    (
+        ByteSink { channel: channel.clone() },
+        ByteStream { channel },
+    )
+}
+
+// A mock sink that implements the Sink trait.
+pub struct ByteSink {
+    channel: Arc<Mutex<ByteChannel>>,
+}
+
+impl Sink for ByteSink {
+    // Writes the message to the buffer.
+    // Resolves the oneshot receiver if-and-only-if the buffer is large enough.
+    async fn send(&mut self, msg: &[u8]) -> Result<(), Error> {
+        let (os_send, data) = {
+            let mut channel = self.channel.lock().unwrap();
+            channel.buffer.extend(msg);
+
+            // If there is a waiter and the buffer is large enough,
+            // return the waiter (while clearing the waiter field).
+            // Otherwise, return early.
+            if channel.waiter.as_ref().map_or(false, |(min, _)| *min <= channel.buffer.len()) {
```

Comment:
nit: use `requested` rather than `min` (this is just because we are returning exactly that amount over the channel)